File: 2021/Day5/Set Matrix zeroes/set_matrix_zeroes.py
class Solution: #By CyFun
    def setZeroes(self, matrix: List[List[int]]) -> None:
        n = len(matrix)
        m = len(matrix[0])
        clear_row = False
        clear_col = False
        i = 0
        j = 0
        while j < m:
            if matrix[0][j] == 0:
                clear_row = True
                break
            j += 1
        while i < n:
            if matrix[i][0] == 0:
                clear_col = True
                break
            i += 1
        i = 1
        while i < n:
            j = 1
            while j < m:
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0
                j += 1
            i += 1
        i = 1
        while i < n:
            if matrix[i][0] == 0:
                j = 1
                while j < m:
                    matrix[i][j] = 0
                    j += 1
            if clear_col:
                matrix[i][0] = 0
            i += 1
        j = 1
        while j < m:
            if matrix[0][j] == 0:
                i = 1
                while i < n:
                    matrix[i][j] = 0
                    i += 1
            if clear_row:
                matrix[0][j] = 0
            j += 1
        if clear_row:
            matrix[0][0] = 0
        if clear_col:
            matrix[0][0] = 0

File: 2021/Day5/Set Matrix zeroes/set_matrix_zeroes.cpp
class Solution
{
public:
    void setZeroes(vector<vector<int>> &matrix)
    {
        //By CyFun
        // Have to check the marks of first row and column before we modify them.
        bool zero_first_row = false;
        bool zero_first_col = false;
        for (int i = 0; i < matrix.size(); i++)
        {
            if (matrix[i][0] == 0)
            {
                zero_first_col = true;
                break;
            }
        }

        for (int j = 0; j < matrix[0].size(); j++)
        {
            if (matrix[0][j] == 0)
            {
                zero_first_row = true;
                break;
            }
        }

        // Scan row and column and set marks
        for (int i = 1; i < matrix.size(); i++)
        {
            for (int j = 1; j < matrix[i].size(); j++)
            {
                if (matrix[i][j] == 0)
                {
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }

        // Set 0 for rows
        for (int i = 1; i < matrix.size(); i++)
        {
            if (matrix[i][0] == 0)
            {
                for (int j = 1; j < matrix[i].size(); j++)
                {
                    matrix[i][j] = 0;
                }
            }
        }

        // Set 0 for columns
        for (int j = 1; j < matrix[0].size(); j++)
        {
            if (matrix[0][j] == 0)
            {
                for (int i = 1; i < matrix.size(); i++)
                {
                    matrix[i][j] = 0;
                }
            }
        }

        // Set 0 for first row.
        if (zero_first_row)
        {
            for (int j = 0; j < matrix[0].size(); j++)
            {
                matrix[0][j] = 0;
            }
        }

        // Set 0 for first column
        if (zero_first_col)
        {
            for (int i = 0; i < matrix.size(); i++)
            {
                matrix[i][0] = 0;
            }
        }
    }
};